Include databricks-agents packages in the mlflow.genai namespace

smoorjani opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/smoorjani/mlflow/pull/15763?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/15763/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/15763/merge#subdirectory=skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/15763/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

<!-- Please fill in changes proposed in this PR. -->
This PR adds a shim for Databricks Agents APIs to be included under the `mlflow.genai` namespace. Specifically, the following namespaces are added:

- `mlflow.genai.datasets` for Managed Datasets (see [docs](https://docs.databricks.com/aws/en/generative-ai/agent-evaluation/review-app#datasets))
- `mlflow.genai.labeling` for Labeling Sessions (see [docs](https://docs.databricks.com/aws/en/generative-ai/agent-evaluation/review-app#labeling-sessions))
- `mflow.genai.label_schemas` for Label Schemas (see [docs](https://docs.databricks.com/aws/en/generative-ai/agent-evaluation/review-app#labeling-sessions))

We expose all methods in the first two through `mlflow.genai` as well. 

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

These APIs are tested by the `databricks-agents` package. We test the imports here:

#### Existing imports work
<img width="477" alt="Screenshot 2025-05-20 at 4 09 45 PM" src="https://github.com/user-attachments/assets/aed1c59d-634a-4d79-8bca-ae6810f7cb11" />


#### Importing without `databricks-agents` installed
<img width="953" alt="Screenshot 2025-05-20 at 4 04 57 PM" src="https://github.com/user-attachments/assets/4bd753e0-488a-4316-bad6-bb728ef3e681" />

For managed datasets, we allow users to import, but fail when the method is called
<img width="952" alt="Screenshot 2025-05-20 at 4 06 02 PM" src="https://github.com/user-attachments/assets/dae63125-7ffd-41e1-ae7d-f49430ad43ff" />

#### Importing with `databricks-agents` installed
<img width="520" alt="Screenshot 2025-05-20 at 4 24 28 PM" src="https://github.com/user-attachments/assets/3e5de1ed-9e1f-4692-b913-6e9f6de5b281" />

Note this calls into the actual code:
<img width="956" alt="Screenshot 2025-05-20 at 4 24 59 PM" src="https://github.com/user-attachments/assets/4f3c10e4-d365-4a0d-8eee-a0d9c280d3f5" />


### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

**TODO**

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

Include `databricks-agents` APIs for datasets, labeling, and judges under the `mlflow.genai` namespace.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [x] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

smoorjani committed:
Include databricks-agents packages in the mlflow.genai namespace

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
Rename review app to labeling

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

dbczumar commented on mlflow/genai/judges/__init__.py:
```diff
@@ -0,0 +1,68 @@
+"""
+Databricks Agent Eval Built-in Judges. For more details see Databricks Agent Evaluation:
+<https://docs.databricks.com/en/generative-ai/agent-evaluation/index.html>
+"""
+
+from mlflow.utils.annotations import experimental
+
+try:
+    from databricks.agents.evals.judges import (
+        chunk_relevance as _chunk_relevance,
+        context_sufficiency as _context_sufficiency,
+        correctness as _correctness,
+        groundedness as _groundedness,
+        guideline_adherence as _guideline_adherence,
+        relevance_to_query as _relevance_to_query,
+        safety as _safety,
+    )
+except ImportError:
+    raise ImportError(
+        "The `databricks-agents` package is required to use mlflow.genai.judges. "
+        "Please install it with `pip install databricks-agents`."
+    )
+
```
We're already importing these in `mlflow.genai.scorers`. I don't think we need to reimport here. cc @B-Step62

B-Step62 replied:
+1, they support direct call as well so users shouldn't need to interact with these.

smoorjani replied:
Ok, nice! Can we also make sure they are marked as experimental?

B-Step62 replied:
Absolutely, it's on me marking all APIs experimental. Will file a PR for it.

dbczumar commented on mlflow/genai/datasets/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+__all__ = [
+    "create_dataset",
+    "Dataset",
```
Shall we call this `EvaluationDataset` to disambiguate it from the other `Dataset` objects in MLflow?

Can we also ensure that this inherits from `mlflow.data.Dataset`?

nsthorat replied:
+1

dbczumar replied:
Sorry, `mlflow.entities.Dataset` is correct. Thanks!

dbczumar replied:
@smoorjani can we update this to inherit from `mlflow.data.Dataset`?

dbczumar replied:
@smoorjani Can we inherit from `mlflow.data.Dataset`? It's fine if we define a new class in MLflow that does some wrapping. Inheriting from `mlflow.data.Dataset` is the necessary approach for compatibility with `mlflow.log_inputs()`, etc.

smoorjani replied:
Yes, confirmed with Daniel, I'll get a PR out for this

B-Step62 commented on mlflow/genai/labeling/__init__.py:
```diff
@@ -0,0 +1,26 @@
+"""
+Databricks Agent Labeling Python SDK. For more details see Databricks Agent Evaluation:
+<https://docs.databricks.com/en/generative-ai/agent-evaluation/index.html>
```
nit: Can we add link to the SDK API doc?

smoorjani replied:
Done!

nsthorat commented on mlflow/genai/labeling/__init__.py:
```diff
@@ -0,0 +1,26 @@
+"""
```
I think we're just doing a vanilla wrapping here -- at the bottom of the doc on the labeling session change there is one important thing about doing anonymous labeling schemas. The way we'll have to do this is namespace the labeling schemas with the labeling session when registering it (so we can potentially have a lot of labeling schemas).

cc @dsmilkov

smoorjani committed:
address pr comments

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
lint

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
updated with latest proposal

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
lint

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

dbczumar commented on mlflow/genai/__init__.py:
```diff
@@ -1,4 +1,15 @@
 from mlflow.genai.scorers import Scorer, scorer
 
-__all__ = ["evaluate", "to_predict_fn", "Scorer", "scorer"]
+try:
+    from databricks.agents.review_app import (
+        ReviewApp,
+        get_review_app,
+    )
+except ImportError:
+    raise ImportError(
+        "The `databricks-agents` package is required to use `mlflow.genai.ReviewApp` and `mlflow.genai.get_review_app`. "
+        "Please install it with `pip install databricks-agents`."
+    )
```
@smoorjani @B-Step62 This will break imports of `mlflow.genai` if `databricks-agents` is not installed, which we don't want to do.

Instead of this approach, I'd prefer that we make every databricks-agents API importable, and raise errors when users try to call them without having databricks-agents SDK installed. Yes, this will require wrapper functions, but it's a significantly better devex because users who forget to install databricks-agents SDK don't get "function not found" errors or wonder why their autocomplete is suddenly returning zero results, etc.

@smoorjani Once we make this change, can we import *everything* except the contents of `databricks.agents.review_app.label_schemas` into `mlflow.genai` for improved ease of access?

dbczumar commented on mlflow/genai/datasets/__init__.py:
```diff
@@ -0,0 +1,29 @@
+
+try:
+    from databricks.agents.datasets import (
+        Dataset as EvaluationDataset,
```
@smoorjani In what version of the databricks-agents SDK will this class inherit from `mlflow.data.Dataset`, instead of `mlflow.entities.Dataset`?

smoorjani replied:
it'll be in the next rc/version of `databricks-agents` so either `1.0.0rc1` or `1.0.0`

dbczumar replied:
Thanks!

dbczumar reviewed (approved):
@smoorjani LGTM provisionally, pending addressing of the remaining comments and manual QA (do we have a manual test plan?)

smoorjani committed:
fixes

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
add everything to mlflow.genai

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
.

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani commented:
> @smoorjani LGTM provisionally, pending addressing of the remaining comments and manual QA (do we have a manual test plan?)

Added manual tests to the description (testing section) as well.

dbczumar commented on mlflow/genai/__init__.py:
```diff
@@ -1,4 +1,46 @@
 from mlflow.genai.scorers import Scorer, scorer
 
-__all__ = ["evaluate", "to_predict_fn", "Scorer", "scorer"]
+try:
+    from mlflow.genai.labeling import (
+        Agent,
+        LabelingSession,
+        ReviewApp,
+        get_review_app,
+        create_labeling_session,
+        get_labeling_sessions,
+        get_labeling_session,
+        delete_labeling_session,
+    )
+    # Stick this at the end to avoid unnecessary warnings
+    from mlflow.genai.datasets import (
+        EvaluationDataset,
+        create_dataset,
+        delete_dataset,
+        get_dataset,
+    )
+except ImportError:
+    # Silently fail if the databricks-agents package is not installed
+    pass
```
@smoorjani I tried out the branch, and `mlflow.genai.create_dataset` isn't available in the `mlflow.genai` namespace because the "library not installed" errors from importing things from within `mlflow.genai.labeling` are masked.

We should move `mlflow.genai.datasets` *outside* of the try / except ImportError block

dbczumar replied:
Can we also add test coverage?

smoorjani replied:
Moved the three methods out of the try/except + added test coverage.

smoorjani committed:
add unit test

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
fix warning

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

harupy commented on tests/genai/test_genai.py:
```diff
@@ -0,0 +1,23 @@
+"""This module tests imports in the mlflow.genai module."""
```
```suggestion
```

smoorjani replied:
removed!

dbczumar commented on tests/genai/test_genai.py:
```diff
@@ -0,0 +1,23 @@
+import pytest
+
+
+def test_namespaced_import_raises_when_agents_not_installed():
+    # Ensure that databricks-agents methods renamespaced under mlflow.genai raise an
+    # ImportError when the databricks-agents package is not installed.
+    import mlflow.genai
+    with pytest.raises(ImportError, match="The `databricks-agents` package is required"):
+        mlflow.genai.create_dataset("test_schema")
+
+    with pytest.raises(ImportError, match="The `databricks-agents` package is required"):
+        mlflow.genai.get_dataset("test_schema")
+
+    with pytest.raises(ImportError, match="The `databricks-agents` package is required"):
+        mlflow.genai.delete_dataset("test_schema")
+
+
+def test_namespaced_import_does_not_exist_when_agents_not_installed():
```
@smoorjani @harupy can we test that `mlflow.genai.create_dataset()` / `mlflow.genai.delete_dataset()` / `mlflow.genai.get_dataset` *are* available, even when `databricks-agents` isn't installed?

smoorjani committed:
add docstring for dataset

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
removed unnecessary module docstring

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani commented on mlflow/genai/datasets/__init__.py:
```diff
@@ -0,0 +1,63 @@
+    return delete_dataset(uc_table_name)
+
+
+def get_dataset(uc_table_name: str) -> "EvaluationDataset":
```
added!

harupy commented on mlflow/genai/__init__.py:
```diff
@@ -1,4 +1,52 @@
+    "LabelingSession",
+    "ReviewApp",
+    "get_review_app",
+    "create_labeling_session",
```
* import fails when `create_labeling_session` is unavailable. Can we append it (and other classes/funcions) to `__all__` only when they are importable?

```
from mlflow.genai import *
```

smoorjani replied:
Fixed! not sure if I can add a unit test since wildcards are not allowed in functions.

smoorjani committed:
fix import mlflow.genai.*

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
fixed datasets & added tests

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
fix docs & linter

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
.

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
fix unit tests

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>

smoorjani committed:
.

Signed-off-by: Samraj Moorjani <samraj.moorjani@databricks.com>
